### PR TITLE
[Model]  XGBoost 베이스라인 모델 추가 및 학습 파이프라인 연결

### PR DESCRIPTION
**Pull request by dimolto3**
## 📌 Summary

신용카드 고객 이탈(Churn) 예측을 위한 **XGBoost 하이퍼파라미터 튜닝 및 최종 모델 학습 파이프라인**을 추가합니다. 16분간의 GridSearchCV를 통해 최적의 성능 조합을 도출하고 일반화 성능을 검증하였습니다.

---

## 🔧 Changes

* **Data**
  * 데이터 common 에서 불러
  * 학습/테스트 데이터 분리 시 `stratify=y` 적용으로 클래스 비율 유지

* **Model**
  * XGBoost 최적 하이퍼파라미터 적용 (`n_estimators: 300`, `learning_rate: 0.05` 등)
  * `GridSearchCV`를 활용한 모델 최적화 프로세스 구축

* **Evaluation**
  * 불균형 데이터셋 대응을 위한 **F1-Score (Macro)** 평가 코드 추가
  * Train/Test 점수 차이(Gap) 분석을 통한 과적합 검증 로직 추가
  * 변수 중요도(Feature Importance) 시각화 기능 추가

---

## 📂 Related Issue

Refs #45

---
## 🧪 Test

* [ ] 모델 학습 정상 실행 (GridSearchCV 포함 전체 파이프라인)
* [ ] 예측 결과 및 F1-Score 산출 확인

---
## 📈 Expected Improvement

Baseline 모델(Logistic Regression / RandomForest 등) 대비
성능 향상을 목표로 합니다.

Target Metrics

- Accuracy: **≥ 0.80**
- Recall: **≥ 0.70**
- F1 Score: **≥ 0.78**
- ROC-AUC: **≥ 0.85**
- PR-AUC: **≥ 0.80**

Churn prediction 문제 특성상
Accuracy뿐 아니라 F1 Score와 PR-AUC를 주요 평가 지표로 사용합니다. 

---

**실제 달성 결과 (XGBoost)**

- **Accuracy:** **0.9743** (목표 초과 달성 ✅)
- **Recall (Class 1):** **0.99** (목표 초과 달성 ✅)
- **F1 Score (Macro):** **0.9520** (목표 초과 달성 ✅)
- **과적합 갭 (Gap):** **0.0275** (Train F1: 0.9795 / Test F1: 0.9520)

---

## 📎 Additional Notes

* GridSearchCV 실행 시 약 16분의 시간이 소요되므로 실행 시 참고 부탁드립니다.
* 현재 모델에서 `Total_Trans_Ct`(거래 횟수)가 가장 중요한 변수로 식별되었습니다.
* 과적합 갭이 안정적인 수치임을 확인하였습니다.

---

## ✔️ Checklist

* [ ] 관련 Issue 연결
* [ ] 코드 스타일 확인
* [ ] 불필요한 파일 제거
* [ ] 문서 업데이트

---

## 👀 Reviewer

@Nobrain711

### CODE REVIEW

File: notebooks/YounJeongYeon/EDA_preprocessing.ipynb [hunk 1]
```diff
@@ -113,97 +108,46 @@
      ]
     }
    ],
-   "execution_count": 1
+   "source": [
+    "%pip install numpy\n",
+    "%pip install pandas\n",
+    "%pip install matplotlib\n",
+    "%pip install seaborn"
+   ]
```
Comment (nobrain711) [inline notebooks/YounJeongYeon/EDA_preprocessing.ipynb]: 해당 셀 확인했습니다 👍

현재 환경에서는 필요한 라이브러리가 이미 설치되어 있어서
`%pip install` 부분은 제거해도 정상적으로 동작할 것 같습니다.

노트북 실행 시 불필요한 설치 과정이 줄어들어
전체 실행 흐름도 조금 더 깔끔해질 것 같습니다.
확인 한 번 부탁드립니다 🙂

File: notebooks/YounJeongYeon/XGBoost.ipynb [hunk 2]
```diff
@@ -0,0 +1,3877 @@
...
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "from sklearn.model_selection import train_test_split\n",
+    "from sklearn.metrics import accuracy_score, classification_report, confusion_matrix\n",
+    "from xgboost import XGBClassifier\n",
```
Comment (nobrain711) [inline notebooks/YounJeongYeon/XGBoost.ipynb]: 해당 import 부분 확인했습니다 👍

`train_test_split`, `accuracy_score`, `classification_report`, `confusion_matrix`, `XGBClassifier` 관련 import는
노트북 상단에서 다른 라이브러리 import와 함께 정리해 두면
전체 코드 흐름을 보기 조금 더 편할 것 같습니다.

가능하면 위쪽 import 셀로 이동해서 정리해 주시면 좋을 것 같습니다 🙂

File: notebooks/YounJeongYeon/XGBoost.ipynb [hunk 3]
```diff
@@ -0,0 +1,3877 @@
...
+    "\n",
+    "# 6. 변수 중요도 시각화\n",
+    "plt.figure(figsize=(10, 8))\n",
+    "importance = pd.Series(xgb_model.feature_importances_, index=X.columns).sort_values(ascending=True)\n",
+    "importance.plot(kind='barh', color='teal')\n",
+    "plt.title('XGBoost Feature Importance (Baseline)')\n",
+    "plt.xlabel('Importance Score')\n",
+    "plt.ylabel('Features')\n",
+    "plt.tight_layout()\n",
+    "plt.show()"
```
Comment (nobrain711) [inline notebooks/YounJeongYeon/XGBoost.ipynb]: 해당 셀 확인했습니다 👍

현재 하나의 셀에서 **데이터 로딩 → 데이터 분리 → 모델 학습 → 평가 → 시각화**까지
모두 처리되고 있어서 실행 흐름을 따라가기가 조금 어려울 수 있을 것 같습니다.

노트북 가독성을 위해 **단계별로 셀을 나누어 정리**하면 좋을 것 같습니다.
예를 들어 아래처럼 구분해 볼 수 있을 것 같습니다.

```python
# 1️⃣ 데이터 불러오기
df = fetch_creditcard()
```

```python
# 2️⃣ 독립변수(X) / 종속변수(y) 분리
X = df.drop('churn', axis=1)
y = df['churn']

# 분리 결과 확인
print("X shape:", X.shape)
print("y shape:", y.shape)
print("\nX 상위 5개 행:")
print(X.head())
print("\ny 상위 5개 값:")
print(y.head())
```

```python
# 3️⃣ 학습 / 테스트 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

# 분할된 데이터 출력
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)
```

```python
# 4️⃣ XGBoost 모델 정의 및 학습
xgb_model = XGBClassifier(
    n_estimators=100,
    learning_rate=0.1,
    max_depth=5,
    random_state=42,
    use_label_encoder=False,
    eval_metric='logloss'
)

xgb_model.fit(X_train, y_train)
```

```python
# 6️⃣ 모델 평가
y_pred = xgb_model.predict(X_test)

print("[모델 평가 결과]\n")
print(f"정확도(Accuracy): {accuracy_score(y_test, y_pred):.4f}\n")
print("분류 리포트(Classification Report):")
print(classification_report(y_test, y_pred))
```

```python
# 7️⃣ 변수 중요도 시각화
plt.figure(figsize=(10, 8))
importance = pd.Series(
    xgb_model.feature_importances_,
    index=X.columns
).sort_values(ascending=True)

importance.plot(kind='barh', color='teal')
plt.title('XGBoost Feature Importance (Baseline)')
plt.xlabel('Importance Score')
plt.ylabel('Features')
plt.tight_layout()
plt.show()
```

이렇게 단계별로 분리하면
노트북 실행 흐름과 디버깅이 조금 더 명확해질 것 같습니다.
검토 부탁드립니다 🙂

File: notebooks/YounJeongYeon/XGBoost.ipynb [hunk 4]
```diff
@@ -0,0 +1,3877 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "f9ca6e83",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import numpy as np\n",
+    "import pandas as pd\n",
+    "import matplotlib.pyplot as plt\n",
+    "import seaborn as sns"
```
Comment (nobrain711) [inline notebooks/YounJeongYeon/XGBoost.ipynb]: 해당 부분 확인했습니다 👍

import 셀에 `from common.fetch_creditcard import fetch_creditcard`를 추가해 주시면 좋을 것 같습니다.

이후 데이터 로딩 부분은 다음처럼 사용할 수 있을 것 같습니다.

```python
df = fetch_creditcard()
```

이렇게 공통 모듈을 사용하면
프로젝트 내 다른 노트북들과도 데이터 로딩 방식이 일관되게 유지될 것 같습니다.
확인 부탁드립니다 🙂

File: notebooks/YounJeongYeon/XGBoost.ipynb [hunk 5]
```diff
@@ -0,0 +1,3877 @@
...
+      "  warnings.warn(smsg, UserWarning)\n",
+      "/usr/local/lib/python3.12/site-packages/xgboost/core.py:158: UserWarning: [08:07:21] WARNING: /workspace/src/learner.cc:740: \n",
+      "Parameters: { \"use_label_encoder\" } are not used.\n",
+      "\n",
+      "  warnings.warn(smsg, UserWarning)\n",
+      "/usr/local/lib/python3.12/site-packages/xgboost/core.py:158: UserWarning: [08:07:23] WARNING: /workspace/src/learner.cc:740: \n",
+      "Parameters: { \"use_label_encoder\" } are not used.\n",
+      "\n",
+      "  warnings.warn(smsg, UserWarning)\n",
+      "/usr/local/lib/python3.12/site-packages/xgboost/core.py:158: UserWarning: [08:07:23] WARNING: /workspace/src/lea
```
Comment (nobrain711) [inline notebooks/YounJeongYeon/XGBoost.ipynb]: 해당 셀 확인했습니다 👍

몇 가지 정리하면 좋을 것 같은 부분이 있어 코멘트 남깁니다.

1️⃣ `pandas`, `numpy`, `matplotlib`, `sklearn.metrics`, `XGBClassifier` import는
이미 노트북 상단 import 셀에서 사용하고 있다면 **중복으로 다시 import하지 않아도 될 것 같습니다.**
가능하면 **상단 import 셀에서 한 번만 관리**하는 방식으로 정리해 주시면 좋을 것 같습니다.

2️⃣ 현재 셀에서 **파라미터 설정 → 모델 학습 → 과적합 검증 → 변수 중요도 시각화 → 임계값 조정 테스트**까지
여러 단계가 한 번에 실행되고 있어서 실행 흐름을 따라가기가 조금 어려울 수 있습니다.

노트북 가독성을 위해 **단계별로 셀을 나누어 정리**해 보면 좋을 것 같습니다.
예를 들어 아래와 같이 구분할 수 있을 것 같습니다.

```python
# 1️⃣ 최적 파라미터 정의
best_params = {
    'colsample_bytree': 1.0,
    'gamma': 0,
    'learning_rate': 0.05,
    'max_depth': 4,
    'n_estimators': 500,
    'scale_pos_weight': 1,
    'subsample': 0.8
}
```

```python
# 2️⃣ 모델 생성 및 학습
best_xgb = XGBClassifier(
    **best_params,
    random_state=42,
    use_label_encoder=False,
    eval_metric='logloss'
)

best_xgb.fit(X_train, y_train)
```

```python
# 3️⃣ Train vs Test 성능 비교 (과적합 확인)
print("=== 1. Train vs Test 점수 비교 ===")

y_train_pred = best_xgb.predict(X_train)
y_test_pred = best_xgb.predict(X_test)

train_f1 = f1_score(y_train, y_train_pred, average='macro')
test_f1 = f1_score(y_test, y_test_pred, average='macro')

print(f"Train F1-Score: {train_f1:.4f}")
print(f"Test F1-Score:  {test_f1:.4f}")
```

```python
# 4️⃣ 변수 중요도 시각화
print("=== 2. 상위 10개 변수 중요도 시각화 ===")

plt.figure(figsize=(10, 6))

importances = pd.Series(
    best_xgb.feature_importances_,
    index=X_train.columns
)

importances.sort_values(ascending=True).tail(10).plot(
    kind='barh',
    color='teal'
)

plt.title('Top 10 Feature Importances in XGBoost')
plt.xlabel('Importance Score')
plt.ylabel('Features')
plt.tight_layout()
plt.show()
```

```python
# 5️⃣ 임계값(Threshold) 조정 테스트
print("=== 3. 임계값(Threshold) 하향 조정 테스트 ===")

churn_probs = best_xgb.predict_proba(X_test)[:, 0]

custom_threshold = 0.3
y_pred_custom = np.where(churn_probs >= custom_threshold, 0, 1)

print(f"[임계값 {custom_threshold} 적용 시 새로운 분류 리포트]")
print(classification_report(y_test, y_pred_custom))
```

이렇게 단계별로 나누면
각 분석 단계(모델 학습 → 과적합 확인 → 변수 중요도 분석 → 임계값 조정)의 목적이 조금 더 명확하게 보이고,
중간 결과 확인이나 디버깅도 수월해질 것 같습니다.

검토 부탁드립니다 🙂
